Editor/Navigator: New - Use JS api commands. [ED-2244]

iNewLegend opened the pull request.

iNewLegend committed:
Editor/Navigator: New - Use JS api commands.

ref: ED-2244

matipojo commented on assets/dev/js/editor/regions/navigator/elements/commands/base/command-navigator-visibility.js:
```diff
@@ -0,0 +1,26 @@
+		containers.forEach( ( container ) => {
+			const { view, model } = container.navigator;
+
+			model.set( 'hidden', this.getVisibility() );
```
```suggestion
			model.set( 'hidden', this.shouldHide() );
```

iNewLegend replied:
Handled: https://github.com/elementor/elementor/pull/14264/commits/b2be23d45787ac398a1243da5b24230a36c44e0e

matipojo commented on assets/dev/js/editor/regions/navigator/elements/commands/base/command-navigator-visibility.js:
```diff
@@ -0,0 +1,26 @@
+	 *
+	 * @returns {Boolean}
+	 */
+	getVisibility() {
```
```suggestion
	shouldHide() {
```

iNewLegend replied:
Handled: https://github.com/elementor/elementor/pull/14264/commits/b2be23d45787ac398a1243da5b24230a36c44e0e

matipojo commented on assets/dev/js/editor/regions/navigator/elements/hooks/ui/document/elements/settings/navigator-render-indicators.js:
```diff
@@ -0,0 +1,38 @@
+	}
+
+	getConditions( args ) {
+		return 'string' === typeof args.settings._title;
```
Check if the navigator is open.

iNewLegend replied:
Handled: https://github.com/elementor/elementor/pull/14264/commits/b2be23d45787ac398a1243da5b24230a36c44e0e

matipojo commented on assets/dev/js/editor/regions/navigator/elements/commands/base/command-navigator.js:
```diff
@@ -0,0 +1,17 @@
+		this.requireContainer( args );
+
+		const { containers = [ args.container ] } = args;
+
```
Check also if the navigator is open.

iNewLegend replied:
Handled: https://github.com/elementor/elementor/pull/14264/commits/b2be23d45787ac398a1243da5b24230a36c44e0e

matipojo commented on assets/dev/js/editor/regions/navigator/elements/hooks/ui/panel/editor/open/navigator-toggle-list.js:
```diff
@@ -0,0 +1,34 @@
+import After from 'elementor-api/modules/hooks/ui/after';
+
+export class NavigatorToggleList extends After {
```
Explain + check if the navigator is open.

iNewLegend replied:
Handled: https://github.com/elementor/elementor/pull/14264/commits/b2be23d45787ac398a1243da5b24230a36c44e0e

matipojo commented on tests/qunit/index.html:
```diff
@@ -2,10 +2,62 @@
 	<span>{{{  }}}</span>
 </script>
 
+
```
Try to use the original templates, maybe via a grunt task that extracts them before the test.

iNewLegend replied:
We cannot yet, its better to have a new task for handling the whole problem for `navigator` As example we need `elementor-editor-wrapper` which created in PHP, the same is also for `elementor-container` which causes issue with `resizeColumn` test

matipojo commented on assets/dev/scss/frontend/_lightbox.scss:
```diff
@@ -136,10 +136,6 @@
 		}
 
 		@media (max-width: $screen-sm-max) {
-			&:focus {
```
Why?

iNewLegend replied:
Handled: https://github.com/elementor/elementor/pull/14264/commits/b2be23d45787ac398a1243da5b24230a36c44e0e

matipojo commented on tests/qunit/setup/editor.js:
```diff
@@ -23,9 +23,15 @@ Marionette.Region.prototype.attachHtml = () => {
-	);
+	const $currentEl = jQuery( containerView.el );
+
+	if ( ! containerView._parent?.$el ) {
```
Explain.

iNewLegend replied:
Handled.

matipojo reviewed (changes requested):
Nice!
Here is the first part of the CR.

iNewLegend committed:
Merge branch 'release/3.2.0' into feature/navigator-to-js-api

iNewLegend committed:
Editor/Navigator: New - Use JS api commands. (R1-Part1)

1.`navigator/elements` hooks and command are should be available only while navigator is open.
2. make more bases from common logic.
3. Navigator toggle list hook didn't work on creation of new widget.
ref: ED-2244

iNewLegend commented on assets/dev/js/editor/regions/navigator/commands/close.js:
```diff
@@ -2,7 +2,13 @@ import CommandBase from 'elementor-api/modules/command-base';
+		}
+
+		this.component.manager.close();
+
```
Handled: https://github.com/elementor/elementor/pull/14264/commits/b2be23d45787ac398a1243da5b24230a36c44e0e

iNewLegend committed:
Editor/Navigator: New - Use JS api commands. (R1N1)

iNewLegend committed:
Merge remote-tracking branch 'Elementor/release/3.2.0' into feature/navigator-to-js-api

ItzhakAvm commented on assets/dev/js/editor/regions/navigator/elements/hooks/ui/base/base.js:
```diff
@@ -0,0 +1,7 @@
+
+export default class Base extends After {
+	getConditions() {
+		return $e.components.get( 'navigator' ).isOpen;
```
The UX ground rules are that each time the user drops an element somewhere, all the tree list down to this element is toggled so the navigator can display the element right away. Considering this condition, when the navigator is open, it should work, the tree list down to the newly dropped element is toggled. But when the navigator is closed, it won't happen. Why is this important? Because of the edge case in which the user opens the navigator after such a drag&drop action. Generally, each navigator change has to happen even though the navigator is closed, so it will be ready to be displayed when the user pop up the navigator again.

matipojo replied:
On open, it will re-render itself.

YitzhakAvm reviewed (commented):
I think the Layout view could of stay the way it is, except of changing function calls to command calls. This way, the commands themselves could call methods from the Layout view, instead of handling the UI changes themselves. I mean.. Generally UI manipulation should happen in the view.

matipojo commented on assets/dev/js/editor/regions/navigator/commands/dock.js:
```diff
@@ -0,0 +1,40 @@
+		}
+
+		// TODO: Move to UI hook.
+		elementorCommon.elements.$body.addClass( 'elementor-navigator-docked' );
```
// TODO: Use the new uiState manager

matipojo commented on assets/dev/js/editor/regions/navigator/commands/dock.js:
```diff
@@ -0,0 +1,40 @@
+		// TODO: Move to UI hook.
+		elementorCommon.elements.$body.addClass( 'elementor-navigator-docked' );
+
+		elementor.navigator.setSize();
```
Use internal command.

matipojo commented on assets/dev/js/editor/regions/navigator/commands/dock.js:
```diff
@@ -0,0 +1,40 @@
+
+		elementor.navigator.setSize();
+
+		const resizableOptions = elementor.navigator.getResizableOptions();
```
```suggestion
		const resizableOptions = this.component.getResizableOptions();
```

matipojo commented on assets/dev/js/editor/regions/navigator/commands/dock.js:
```diff
@@ -0,0 +1,40 @@
+
+		const resizableOptions = elementor.navigator.getResizableOptions();
+
+		elementor.navigator.$el.css( {
```
Replace all `elementor.navigator` with `this.component` + BC

matipojo commented on assets/dev/js/editor/regions/navigator/commands/dock.js:
```diff
@@ -0,0 +1,40 @@
+			elementor.navigator.$el.resizable( 'destroy' );
+		}
+
+		resizableOptions.handles = elementorCommon.config.isRTL ? 'e' : 'w';
```
Move to `getResizableOptions`

matipojo commented on assets/dev/js/editor/regions/navigator/commands/undock.js:
```diff
@@ -0,0 +1,34 @@
+import CommandBase from 'elementor-api/modules/command-base';
+
+export class Undock extends CommandBase {
+	apply( args ) {
```
See the `dock` comments

matipojo commented on assets/dev/js/editor/regions/navigator/element-empty.js:
```diff
@@ -1,13 +0,0 @@
-export default class extends Marionette.ItemView {
```
Restore.

matipojo commented on assets/dev/js/editor/regions/navigator/element-empty.js:
```diff
@@ -1,13 +0,0 @@
-		return 'elementor-empty-view';
-	}
-
-	onRendr() {
```
Fix

matipojo commented on assets/dev/js/editor/regions/navigator/elements/commands/base/command-navigator-visibility.js:
```diff
@@ -0,0 +1,26 @@
+import CommandNavigator from './command-navigator';
```
Rename  to `show-hide`

matipojo commented on assets/dev/js/editor/regions/navigator/elements/commands/base/command-navigator.js:
```diff
@@ -0,0 +1,21 @@
+export class CommandNavigator extends CommandBase {
+	validateArgs( args ) {
+		if ( ! $e.components.get( 'navigator' ).isOpen ) {
+			throw Error( `Cannot use: '${ this.component.getNamespace() }' while navigator is closed.` );
```
Devtools warning

matipojo commented on assets/dev/js/editor/regions/navigator/elements/commands/base/command-navigator.js:
```diff
@@ -0,0 +1,21 @@
+			throw Error( `Cannot use: '${ this.component.getNamespace() }' while navigator is closed.` );
+		}
+
+		this.requireContainer( args );
```
Check why it's needed and how it's possible it works?

matipojo commented on assets/dev/js/editor/regions/navigator/elements/commands/hide.js:
```diff
@@ -0,0 +1,9 @@
+import CommandNavigatorVisibility from './base/command-navigator-visibility';
+
+export class Hide extends CommandNavigatorVisibility {
+	shouldHide() {
```
Find a better solution (that will fit also the enable-disable command base)

matipojo commented on assets/dev/js/editor/regions/navigator/elements/hooks/ui/document/elements/settings/navigator-render-indicators.js:
```diff
@@ -0,0 +1,38 @@
+				if ( Object.keys( container.settings.changed ).filter( ( key ) => indicatorSettings.settingKeys.includes( key ) ).length ) {
+					view.renderIndicators();
+
+					return false;
```
Try remove.

matipojo commented on assets/dev/js/editor/regions/navigator/elements/hooks/ui/document/elements/settings/navigator-render-indicators.js:
```diff
@@ -0,0 +1,38 @@
+		containers.forEach( ( container ) => {
+			const { view, model } = container.navigator;
+
+			// Prevent empty title. ( consider move to another hook ).
+			if ( 0 === args.settings._title.length ) {
+				view.ui.title.text( model.getTitle() );
+			}
```
Explain

matipojo commented on assets/dev/js/editor/regions/navigator/elements/hooks/ui/document/elements/settings/navigator-render-indicators.js:
```diff
@@ -0,0 +1,38 @@
+	}
+
+	getConditions( args ) {
+		return super.getConditions() && 'string' === typeof args.settings._title;
```
Why?

matipojo commented on assets/dev/js/editor/regions/navigator/elements/hooks/ui/base/base-hide-show.js:
```diff
@@ -0,0 +1,9 @@
+import Base from './base';
```
Move all show/hide functionality to the commands themselves.

iNewLegend commented:
Maybe move the view to React or WebComponent.

iNewLegend committed:
Merge remote-tracking branch 'Elementor/release/3.5.0' into feature/navigator-to-js-api

iNewLegend committed:
Merge remote-tracking branch 'Elementor/release/3.5.0' into feature/navigator-to-js-api

iNewLegend committed:
Navigator: Move setSize, to internal command.
